libkbfs/mdserver_disk.go

```diff
@@ -5,97 +5,66 @@
 package libkbfs
 
 import (
-	"bytes"
-	"encoding/binary"
 	"errors"
-	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"reflect"
 	"sync"
-	"time"
 
 	"github.com/keybase/client/go/logger"
 	keybase1 "github.com/keybase/client/go/protocol"
 	"github.com/syndtr/goleveldb/leveldb"
 	"github.com/syndtr/goleveldb/leveldb/storage"
-	"github.com/syndtr/goleveldb/leveldb/util"
 	"golang.org/x/net/context"
 )
 
-// TODO: Convert this to use flat files and a journal.
+type mdServerDiskShared struct {
+	dirPath string
 
-// MDServerDisk just stores blocks in local leveldb instances.
-type MDServerDisk struct {
-	config   Config
-	handleDb *leveldb.DB // folder handle                  -> folderId
-	mdDb     *leveldb.DB // folderId+[branchId]+[revision] -> mdBlockLocal
-	branchDb *leveldb.DB // folderId+deviceKID             -> branchId
-	log      logger.Logger
-
-	locksMutex *sync.Mutex
-	locksDb    *leveldb.DB // folderId -> deviceKID
+	// Protects handleDb and tlfStorage. After Shutdown() is
+	// called, both handleDb and tlfStorage are nil.
+	lock       sync.RWMutex
+	handleDb   *leveldb.DB // folder handle -> folderId
+	tlfStorage map[TlfID]*mdServerTlfStorage
 
 	updateManager *mdServerLocalUpdateManager
 
-	shutdownLock *sync.RWMutex
-	shutdown     *bool
 	shutdownFunc func(logger.Logger)
 }
 
-var _ mdServerLocal = (*MDServerDisk)(nil)
+// MDServerDisk stores all info on disk.
+type MDServerDisk struct {
+	codec  Codec
```

Comment:
When there's more than one of these config-originating objects, I'm not a big fan of storing them individually.  We have some tests that swap things into the `Config` object on the fly, and I can imagine a time when that happens in production too, especially for things like crypto and kbpki.  Plus it seems more difficult for the test to construct four separate things, rather than just using a standard test config.  Can we change it back? Or am I missing something about why this is better?